File: src/main/java/org/jcodec/containers/mp4/demuxer/DashStreamDemuxer.java
package org.jcodec.containers.mp4.demuxer;

import java.io.Closeable;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadFactory;

import org.jcodec.api.JCodecException;
import org.jcodec.common.Codec;
import org.jcodec.common.Demuxer;
import org.jcodec.common.DemuxerTrack;
import org.jcodec.common.DemuxerTrackMeta;
import org.jcodec.common.SeekableDemuxerTrack;
import org.jcodec.common.TrackType;
import org.jcodec.common.io.NIOUtils;
import org.jcodec.common.model.Packet;
import org.jcodec.containers.mp4.MP4Packet;
import org.jcodec.containers.mp4.MPDModel;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Demuxes one track out of multiple DASH fragments
 * 
 * @author The JCodec project
 * 
 */
public class DashStreamDemuxer implements Demuxer {
    private List<DashStreamDemuxerTrack> tracks;
    private List<SeekableDemuxerTrack> coded;
    private URL url;
    private MPDModel.MPD mpd;

    public static int INIT_SIZE = 3;

    public DashStreamDemuxer(URL url) throws IOException {
        this.url = url;
        this.tracks = new LinkedList<DashStreamDemuxer.DashStreamDemuxerTrack>();
        this.coded = new LinkedList<SeekableDemuxerTrack>();
        try {
            this.mpd = MPDModel.parse(url);
            if (this.mpd != null && this.mpd.periods != null && this.mpd.periods.size() > 0) {
                MPDModel.Period period = this.mpd.periods.get(0);
                for (MPDModel.AdaptationSet adaptationSet : period.adaptationSets) {
                    DashStreamDemuxerTrack tr = new DashStreamDemuxerTrack(url, adaptationSet, period);
                    tracks.add(tr);
                    coded.add(new CodecMP4DemuxerTrack(tr));
                }
            }
        } catch (JCodecException e) {
            throw new IOException(e);
        }
    }

    static class DashStreamDemuxerTrack implements SeekableDemuxerTrack, Closeable {
        private MPDModel.AdaptationSet adaptationSet;
        private URL url;
        private File initFile;
        private String selectedRprz;
        private Map<Integer, Future<DashMP4DemuxerTrack>> fragments = new HashMap<Integer, Future<DashMP4DemuxerTrack>>();
        private int curFragNo;
        private boolean streaming;
        private int maxDownloadAttampts = 1024;
        private ExecutorService threadPool;
        private int globalFrame;
        private MPDModel.Period period;
        private int frameRate;
        private double segmentDuration;
        private int[] seekFrames;
        static int next_id = 0;
        int id;

        public DashStreamDemuxerTrack(URL url, MPDModel.AdaptationSet adaptationSet, MPDModel.Period period)
                throws IOException {
            this.url = url;
            this.period = period;
            this.adaptationSet = adaptationSet;
            this.threadPool = Executors.newFixedThreadPool(INIT_SIZE, new ThreadFactory() {
                public Thread newThread(Runnable r) {
                    Thread t = Executors.defaultThreadFactory().newThread(r);
                    t.setDaemon(true);
                    return t;
                }
            });
            if (adaptationSet.representations.size() > 0) {
                MPDModel.Representation rprz = adaptationSet.representations.get(0);
                selectedRprz = rprz.id;
                frameRate = rprz.frameRate;
                MPDModel.SegmentTemplate stpl = getSegmentTemplate();
                segmentDuration = stpl != null ? ((double) stpl.duration) / stpl.timescale : period.duration.sec;

                downloadInit();
                for (int i = 0; i < INIT_SIZE; i++) {
                    scheduleFragment(i);
                }
                int numSeg = (int) (period.duration.sec / segmentDuration);
                int segmentFrames = (int) (segmentDuration * frameRate);
                seekFrames = new int[numSeg];
                for (int i = 0, tmp = 0; i < numSeg; i++, tmp += segmentFrames) {
                    seekFrames[i] = tmp + 1;
                }
            }
            id = next_id++;
        }

        private MPDModel.Representation getRrprz() {
            for (MPDModel.Representation rprz : adaptationSet.representations) {
                if (rprz.id.equals(selectedRprz))
                    return rprz;
            }
            return null;
        }

        private MPDModel.SegmentTemplate getSegmentTemplate() {
            MPDModel.Representation rprz = getRrprz();
            return adaptationSet.segmentTemplate == null ? rprz.segmentTemplate : adaptationSet.segmentTemplate;
        }

        private static void sleepQuiet(int millis) {
            try {
                Thread.sleep(millis);
            } catch (InterruptedException e) {
            }
        }

        private DashMP4DemuxerTrack downloadFrag(int no) throws IOException {
            MPDModel.Representation rprz = getRrprz();
            MPDModel.SegmentTemplate stpl = getSegmentTemplate();
            for (int i = 0; i < maxDownloadAttampts; i++) {
                try {
                    URL urlInit = null;
                    if (stpl != null && stpl.media != null) {
                        Map<String, Object> vals = new HashMap<String, Object>();
                        vals.put("RepresentationID", selectedRprz);
                        vals.put("Number", stpl.startNumber + no);
                        String tmp = fillTemplate(stpl.media, vals);

                        urlInit = new URL(url, tmp);
                    } else if (rprz.segmentList != null && rprz.segmentList.segmentUrls.size() > no) {
                        MPDModel.SegmentURL segmentURL = rprz.segmentList.segmentUrls.get(no);
                        urlInit = new URL(url, segmentURL.media);
                    }
                    if (urlInit != null) {
                        File tempFile = File.createTempFile("org.jcodec", fileName(urlInit.getPath()));
                        System.out.println("Fetching fragment: " + urlInit.toExternalForm());
                        NIOUtils.fetchUrl(urlInit, tempFile);
                        DashMP4DemuxerTrack demuxer = DashMP4DemuxerTrack
                                .createFromFiles(Arrays.asList(new File[] { initFile, tempFile }));
                        demuxer.setDurationHint(segmentDuration);
                        return demuxer;
                    } else if (rprz.baseURL != null) {
                        DashMP4DemuxerTrack demuxer = DashMP4DemuxerTrack
                                .createFromFiles(Arrays.asList(new File[] { initFile, initFile }));
                        return demuxer;
                    }
                    break;
                } catch (FileNotFoundException e) {
                    if (!streaming)
                        return null;
                    sleepQuiet(100);
                }
            }
            return null;
        }

        private String fillTemplate(String media, Map<String, Object> vals) {
            StringBuilder builder = new StringBuilder();
            char[] charArray = media.toCharArray();
            int varStart = 0;
            for (int i = 0; i < charArray.length; i++) {
                if ('$' == charArray[i]) {
                    if (varStart == 0) {
                        varStart = i + 1;
                    } else {
                        String var = new String(charArray, varStart, i - varStart);
                        int formatStart = var.indexOf('%');
                        if (formatStart != -1) {
                            String format = var.substring(formatStart);
                            var = var.substring(0, formatStart);
                            Object object = vals.get(var);
                            if (object != null) {
                                String val = String.format(format, object);
                                builder.append(val);
                            }
                        } else {
                            Object object = vals.get(var);
                            if (object != null) {
                                builder.append(String.valueOf(object));
                            }
                        }
                        varStart = 0;
                    }
                } else if (varStart == 0) {
                    builder.append(charArray[i]);
                }
            }
            return builder.toString();
        }

        private String fileName(String path) {
            String[] split = path.split("/");
            return split[split.length - 1];
        }

        private void downloadInit() throws IOException {
            MPDModel.Representation rprz = getRrprz();
            if (getSegmentTemplate() != null && getSegmentTemplate().initialization != null) {
                String tmp = getSegmentTemplate().initialization.replace("$RepresentationID$", selectedRprz);
                URL urlInit = new URL(url, tmp);
                File tempFile = File.createTempFile("org.jcodec", fileName(urlInit.getPath()));
                System.out.println("Fetching init: " + urlInit.toExternalForm());
                NIOUtils.fetchUrl(urlInit, tempFile);
                initFile = tempFile;
            } else if (rprz.baseURL != null) {
                URL urlInit = new URL(url, rprz.baseURL);
                File tempFile = File.createTempFile("org.jcodec", fileName(urlInit.getPath()));
                System.out.println("Fetching init: " + urlInit.toExternalForm());
                NIOUtils.fetchUrl(urlInit, tempFile);
                initFile = tempFile;
            } else if (rprz.segmentList != null && rprz.segmentList.initialization != null) {
                URL urlInit = new URL(url, rprz.segmentList.initialization.sourceURL);
                File tempFile = File.createTempFile("org.jcodec", fileName(urlInit.getPath()));
                System.out.println("Fetching init: " + urlInit.toExternalForm());
                NIOUtils.fetchUrl(urlInit, tempFile);
                initFile = tempFile;
            }
        }

        @Override
        public Packet nextFrame() throws IOException {
            try {
                Future<DashMP4DemuxerTrack> curFrag = fragments.get(curFragNo);
                MP4Packet nextFrame = null;
                if (curFrag != null) {
                    nextFrame = getCurFrag(curFrag).nextFrame();
                    if (nextFrame == null) {
                        getCurFrag(curFrag).close();
                        fragments.put(curFragNo, null);
                        curFragNo++;
                    }
                }
                if (nextFrame != null) {
                    ++globalFrame;
                    return setPts(nextFrame);
                }

                curFrag = fragments.get(curFragNo);

                if (curFrag == null) {
                    for (int i = curFragNo; i < curFragNo + INIT_SIZE; i++)
                        scheduleFragment(curFragNo);
                    curFrag = fragments.get(curFragNo);
                }

                if (curFrag == null)
                    return null;

                ++globalFrame;
                return setPts(getCurFrag(curFrag).nextFrame());
            } catch (ExecutionException e) {
                throw new RuntimeException("Execution problem", e);
            }
        }

        private DashMP4DemuxerTrack getCurFrag(Future<DashMP4DemuxerTrack> curFrag) throws ExecutionException {
            while (true) {
                try {
                    return curFrag.get();
                } catch (InterruptedException e) {
                }
            }
        }

        private MP4Packet setPts(MP4Packet frame) {
            double off = curFragNo * segmentDuration;
            frame.setPts((long) (frame.getPts() + off * frame.getTimescale()));
            frame.setMediaPts((long) (frame.getMediaPts() + off * frame.getTimescale()));
            frame.setFrameNo(globalFrame - 1);
            if (id == 5)
            System.out.println(
                    String.format("[%d] PTS: %f DUR: %s", id, (float) frame.getPtsD(), (float) frame.getDurationD()));
            return frame;
        }

        private void scheduleFragment(final int fragNo) {
            if (fragments.get(fragNo) == null) {
                Future<DashMP4DemuxerTrack> future = threadPool.submit(new Callable<DashMP4DemuxerTrack>() {
                    @Override
                    public DashMP4DemuxerTrack call() throws Exception {
                        return downloadFrag(fragNo);
                    }
                });
                fragments.put(fragNo, future);
            }
        }

        @Override
        public DemuxerTrackMeta getMeta() {
            Future<DashMP4DemuxerTrack> future = fragments.get(curFragNo);
            if (future == null)
                return null;
            try {
                DashMP4DemuxerTrack frag = getCurFrag(future);

                MP4DemuxerTrackMeta fragMeta = (MP4DemuxerTrackMeta) frag.getMeta();
                double totalDuration = period.duration.sec;
                int totalFrames = (int) period.duration.sec * frameRate;
                return new MP4DemuxerTrackMeta(fragMeta.getType(), fragMeta.getCodec(), totalDuration, seekFrames,
                        totalFrames, fragMeta.getCodecPrivate(), fragMeta.getVideoCodecMeta(),
                        fragMeta.getAudioCodecMeta(), fragMeta.getSampleEntries(), fragMeta.getCodecPrivateOpaque());
            } catch (ExecutionException e) {
                throw new RuntimeException("Execution problem", e);
            }
        }

        @Override
        public void close() throws IOException {
            Set<Entry<Integer, Future<DashMP4DemuxerTrack>>> entrySet = fragments.entrySet();
            for (Entry<Integer, Future<DashMP4DemuxerTrack>> entry : entrySet) {
                if (entry.getValue() != null) {
                    try {
                        getCurFrag(entry.getValue()).close();
                    } catch (ExecutionException e) {
                        throw new IOException("Execution problem", e);
                    }
                    entry.setValue(null);
                }
            }
        }

        @Override
        public boolean gotoFrame(long frameNo) throws IOException {
            if (frameNo != 0)
                return false;
            curFragNo = 0;
            globalFrame = 0;
            for (int i = 0; i < INIT_SIZE; i++) {
                scheduleFragment(i);
            }
            return true;
        }

        @Override
        public boolean gotoSyncFrame(long frameNo) throws IOException {
            return false;
        }

        @Override
        public long getCurFrame() {
            return globalFrame;
        }

        @Override
        public void seek(double second) throws IOException {
            throw new RuntimeException("unimpl");
        }
    }

    @Override
    public void close() throws IOException {
        for (DashStreamDemuxerTrack track : tracks) {
            track.close();
        }
    }

    @Override
    public List<? extends DemuxerTrack> getTracks() {
        return coded;
    }

    @Override
    public List<? extends DemuxerTrack> getVideoTracks() {
        ArrayList<SeekableDemuxerTrack> result = new ArrayList<SeekableDemuxerTrack>();
        for (SeekableDemuxerTrack demuxerTrack : coded) {
            DemuxerTrackMeta meta = demuxerTrack.getMeta();
            if (meta.getType() == TrackType.VIDEO)
                result.add(demuxerTrack);
        }
        return result;
    }

    @Override
    public List<? extends DemuxerTrack> getAudioTracks() {
        ArrayList<SeekableDemuxerTrack> result = new ArrayList<SeekableDemuxerTrack>();
        for (SeekableDemuxerTrack demuxerTrack : coded) {
            DemuxerTrackMeta meta = demuxerTrack.getMeta();
            if (meta.getType() == TrackType.AUDIO)
                result.add(demuxerTrack);
        }
        return result;
    }
}

File: src/main/java/org/jcodec/containers/mp4/demuxer/DashStreamDemuxer.kt
package org.jcodec.containers.mp4.demuxer

import org.jcodec.api.JCodecException
import org.jcodec.common.*
import org.jcodec.common.io.NIOUtils
import org.jcodec.common.model.Packet
import org.jcodec.containers.mp4.MP4Packet
import org.jcodec.containers.mp4.MPDModel
import org.jcodec.containers.mp4.MPDModel.AdaptationSet
import org.jcodec.containers.mp4.MPDModel.MPD
import org.jcodec.containers.mp4.demuxer.DashMP4DemuxerTrack.Companion.createFromFiles
import java.io.Closeable
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.net.URL
import java.util.*
import java.util.concurrent.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Demuxes one track out of multiple DASH fragments
 *
 * @author The JCodec project
 */
class DashStreamDemuxer(url: URL) : Demuxer {
    private val tracks: MutableList<DashStreamDemuxerTrack>
    private val coded: MutableList<SeekableDemuxerTrack?>
    private var mpd: MPD? = null

    internal class DashStreamDemuxerTrack(private val url: URL, private val adaptationSet: AdaptationSet, private val period: MPDModel.Period) : SeekableDemuxerTrack, Closeable {
        private var initFile: File? = null
        private var selectedRprz: String? = null
        private val fragments: MutableMap<Int, Future<DashMP4DemuxerTrack?>?> = HashMap()
        private var curFragNo = 0
        private val streaming = false
        private val maxDownloadAttampts = 1024
        private val threadPool: ExecutorService
        private var globalFrame = 0
        private var frameRate = 0f
        private var segmentDuration = 0.0
        private var seekFrames: IntArray = IntArray(0)
        var id: Int
        private val rrprz: MPDModel.Representation?
            private get() {
                for (rprz in adaptationSet.representations) {
                    if (rprz.id == selectedRprz) return rprz
                }
                return null
            }

        private val segmentTemplate: MPDModel.SegmentTemplate?
            private get() {
                val rprz = rrprz
                return if (adaptationSet.segmentTemplate == null) rprz!!.segmentTemplate else adaptationSet.segmentTemplate
            }

        @Throws(IOException::class)
        private fun downloadFrag(no: Int): DashMP4DemuxerTrack? {
            val rprz = rrprz
            val stpl = segmentTemplate
            for (i in 0 until maxDownloadAttampts) {
                try {
                    var urlInit: URL? = null
                    if (stpl != null && stpl.media != null) {
                        val vals: MutableMap<String, Any?> = HashMap()
                        vals["RepresentationID"] = selectedRprz
                        vals["Number"] = stpl.startNumber + no
                        val tmp = fillTemplate(stpl.media, vals)
                        urlInit = URL(url, tmp)
                    } else if (rprz!!.segmentList != null && rprz.segmentList.segmentUrls.size > no) {
                        val segmentURL = rprz.segmentList.segmentUrls[no]
                        urlInit = URL(url, segmentURL.media)
                    }
                    if (urlInit != null) {
                        val tempFile = File.createTempFile("org.jcodec", fileName(urlInit.path))
                        println("Fetching fragment: " + urlInit.toExternalForm())
                        NIOUtils.fetchUrl(urlInit, tempFile)
                        val demuxer = createFromFiles(Arrays.asList(*arrayOf(initFile, tempFile)))
                        demuxer.setDurationHint(segmentDuration)
                        return demuxer
                    } else if (rprz!!.baseURL != null) {
                        return createFromFiles(Arrays.asList(initFile, initFile))
                    }
                    break
                } catch (e: FileNotFoundException) {
                    if (!streaming) return null
                    sleepQuiet(100)
                }
            }
            return null
        }

        private fun fillTemplate(media: String, vals: Map<String, Any?>): String {
            val builder = StringBuilder()
            val charArray = media.toCharArray()
            var varStart = 0
            for (i in charArray.indices) {
                if ('$' == charArray[i]) {
                    if (varStart == 0) {
                        varStart = i + 1
                    } else {
                        var `var` = String(charArray, varStart, i - varStart)
                        val formatStart = `var`.indexOf('%')
                        if (formatStart != -1) {
                            val format = `var`.substring(formatStart)
                            `var` = `var`.substring(0, formatStart)
                            val `object` = vals[`var`]
                            if (`object` != null) {
                                val `val` = String.format(format, `object`)
                                builder.append(`val`)
                            }
                        } else {
                            val `object` = vals[`var`]
                            if (`object` != null) {
                                builder.append(`object`.toString())
                            }
                        }
                        varStart = 0
                    }
                } else if (varStart == 0) {
                    builder.append(charArray[i])
                }
            }
            return builder.toString()
        }

        private fun fileName(path: String): String {
            val split = path.split("/".toRegex()).toTypedArray()
            return split[split.size - 1]
        }

        @Throws(IOException::class)
        private fun downloadInit() {
            val rprz = rrprz
            if (segmentTemplate != null && segmentTemplate!!.initialization != null) {
                val tmp = segmentTemplate!!.initialization.replace("\$RepresentationID$", selectedRprz!!)
                val urlInit = URL(url, tmp)
                val tempFile = File.createTempFile("org.jcodec", fileName(urlInit.path))
                println("Fetching init: " + urlInit.toExternalForm())
                NIOUtils.fetchUrl(urlInit, tempFile)
                initFile = tempFile
            } else if (rprz!!.baseURL != null) {
                val urlInit = URL(url, rprz.baseURL)
                val tempFile = File.createTempFile("org.jcodec", fileName(urlInit.path))
                println("Fetching init: " + urlInit.toExternalForm())
                NIOUtils.fetchUrl(urlInit, tempFile)
                initFile = tempFile
            } else if (rprz.segmentList != null && rprz.segmentList.initialization != null) {
                val urlInit = URL(url, rprz.segmentList.initialization.sourceURL)
                val tempFile = File.createTempFile("org.jcodec", fileName(urlInit.path))
                println("Fetching init: " + urlInit.toExternalForm())
                NIOUtils.fetchUrl(urlInit, tempFile)
                initFile = tempFile
            }
        }

        @Throws(IOException::class)
        override fun nextFrame(): Packet? {
            return try {
                var curFrag = fragments[curFragNo]
                var nextFrame: MP4Packet? = null
                if (curFrag != null) {
                    nextFrame = getCurFrag(curFrag)!!.nextFrame()
                    if (nextFrame == null) {
                        getCurFrag(curFrag)!!.close()
                        fragments[curFragNo] = null
                        curFragNo++
                    }
                }
                if (nextFrame != null) {
                    ++globalFrame
                    return setPts(nextFrame)
                }
                curFrag = fragments[curFragNo]
                if (curFrag == null) {
                    for (i in curFragNo until curFragNo + INIT_SIZE) scheduleFragment(curFragNo)
                    curFrag = fragments[curFragNo]
                }
                if (curFrag == null) return null
                ++globalFrame
                setPts(getCurFrag(curFrag)!!.nextFrame()!!)
            } catch (e: ExecutionException) {
                throw RuntimeException("Execution problem", e)
            }
        }

        @Throws(ExecutionException::class)
        private fun getCurFrag(curFrag: Future<DashMP4DemuxerTrack?>?): DashMP4DemuxerTrack? {
            while (true) {
                try {
                    return curFrag!!.get()
                } catch (e: InterruptedException) {
                }
            }
        }

        private fun setPts(frame: MP4Packet): MP4Packet {
            val off = curFragNo * segmentDuration
            frame.setPts((frame.getPts() + off * frame.getTimescale()).toLong())
            frame.mediaPts = (frame.mediaPts + off * frame.getTimescale()).toLong()
            frame.setFrameNo(globalFrame - 1)
            if (id == 5) println(String.format("[%d] PTS: %f DUR: %s", id, frame.ptsD.toFloat(), frame.durationD.toFloat()))
            return frame
        }

        private fun scheduleFragment(fragNo: Int) {
            if (fragments[fragNo] == null) {
                val future: Future<DashMP4DemuxerTrack?> = threadPool.submit(Callable { downloadFrag(fragNo) })
                fragments[fragNo] = future
            }
        }

        override fun getMeta(): DemuxerTrackMeta? {
            val future = fragments[curFragNo] ?: return null
            return try {
                val frag = getCurFrag(future)!!
                val fragMeta = frag.meta as MP4DemuxerTrackMeta
                val totalDuration = period.duration.sec
                val totalFrames = (period.duration.sec * frameRate).toInt()
                MP4DemuxerTrackMeta(fragMeta.type, fragMeta.codec, totalDuration, seekFrames,
                        totalFrames, fragMeta.codecPrivate, fragMeta.videoCodecMeta,
                        fragMeta.audioCodecMeta, fragMeta.sampleEntries, fragMeta.codecPrivateOpaque)
            } catch (e: ExecutionException) {
                throw RuntimeException("Execution problem", e)
            }
        }

        @Throws(IOException::class)
        override fun close() {
            val entrySet: Set<MutableMap.MutableEntry<Int, Future<DashMP4DemuxerTrack?>?>> = fragments.entries
            for (entry in entrySet) {
                if (entry.value != null) {
                    try {
                        getCurFrag(entry.value)!!.close()
                    } catch (e: ExecutionException) {
                        throw IOException("Execution problem", e)
                    }
                    entry.setValue(null)
                }
            }
        }

        @Throws(IOException::class)
        override fun gotoFrame(frameNo: Long): Boolean {
            if (frameNo != 0L) return false
            curFragNo = 0
            globalFrame = 0
            for (i in 0 until INIT_SIZE) {
                scheduleFragment(i)
            }
            return true
        }

        @Throws(IOException::class)
        override fun gotoSyncFrame(frameNo: Long): Boolean {
            return false
        }

        override fun getCurFrame(): Long {
            return globalFrame.toLong()
        }

        @Throws(IOException::class)
        override fun seek(second: Double) {
            throw RuntimeException("unimpl")
        }

        companion object {
            var next_id = 0
            private fun sleepQuiet(millis: Int) {
                try {
                    Thread.sleep(millis.toLong())
                } catch (e: InterruptedException) {
                }
            }
        }

        init {
            threadPool = Executors.newFixedThreadPool(INIT_SIZE) { r ->
                val t = Executors.defaultThreadFactory().newThread(r)
                t.isDaemon = true
                t
            }
            if (adaptationSet.representations.size > 0) {
                val rprz = adaptationSet.representations[0]
                selectedRprz = rprz.id
                frameRate = if (rprz.frameRate == null) 0f else rprz.frameRate.scalar()
                val stpl = segmentTemplate
                segmentDuration = if (stpl != null) stpl.duration.toDouble() / stpl.timescale else period.duration.sec
                downloadInit()
                for (i in 0 until INIT_SIZE) {
                    scheduleFragment(i)
                }
                val numSeg = (period.duration.sec / segmentDuration).toInt()
                val segmentFrames = (segmentDuration * frameRate).toInt()
                seekFrames = IntArray(numSeg)
                var i = 0
                var tmp = 0
                while (i < numSeg) {
                    seekFrames[i] = tmp + 1
                    i++
                    tmp += segmentFrames
                }
            }
            id = next_id++
        }
    }

    @Throws(IOException::class)
    override fun close() {
        for (track in tracks) {
            track.close()
        }
    }

    override fun getTracks(): List<DemuxerTrack?> {
        return coded
    }

    override fun getVideoTracks(): List<DemuxerTrack?> {
        val result = ArrayList<SeekableDemuxerTrack?>()
        for (demuxerTrack in coded) {
            val meta = demuxerTrack!!.meta
            if (meta.type == TrackType.VIDEO) result.add(demuxerTrack)
        }
        return result
    }

    override fun getAudioTracks(): List<DemuxerTrack?> {
        val result = ArrayList<SeekableDemuxerTrack?>()
        for (demuxerTrack in coded) {
            val meta = demuxerTrack!!.meta
            if (meta.type == TrackType.AUDIO) result.add(demuxerTrack)
        }
        return result
    }

    companion object {
        var INIT_SIZE = 3
    }

    init {
        tracks = LinkedList()
        coded = LinkedList()
        try {
            val _mpd = MPDModel.parse(url)
            if (_mpd != null && _mpd.periods != null && _mpd.periods.size > 0) {
                val period = _mpd.periods[0]
                for (adaptationSet in period.adaptationSets) {
                    val tr = DashStreamDemuxerTrack(url, adaptationSet, period)
                    tracks.add(tr)
                    coded.add(CodecMP4DemuxerTrack(tr))
                }
            }
            mpd = _mpd
        } catch (e: JCodecException) {
            throw IOException(e)
        }
    }
}